devito/petsc/solve.py

```diff
@@ -0,0 +1,116 @@
+from functools import singledispatch
+
+from devito.finite_differences.differentiable import Add, Mul, EvalDerivative
+from devito.finite_differences.derivative import Derivative
+from devito.types import Eq
+from devito.operations.solve import eval_time_derivatives
+from devito.petsc.types import PETScArray, LinearSolveExpr, MatVecEq, RHSEq
+
+from sympy import simplify
+
+
+__all__ = ['PETScSolve']
+
+
+def PETScSolve(eq, target, bcs=None, solver_parameters=None, **kwargs):
+    # TODO: Add check for time dimensions and utilise implicit dimensions.
+
+    is_time_dep = any(dim.is_Time for dim in target.dimensions)
+    # TODO: Current assumption is rhs is part of pde that remains
+    # constant at each timestep. Need to insert function to extract this from eq.
+    y_matvec, x_matvec, b_tmp = [
+        PETScArray(name=f'{prefix}_{target.name}',
+                   dtype=target.dtype,
+                   dimensions=target.space_dimensions,
+                   shape=target.grid.shape,
+                   liveness='eager',
+                   halo=target.halo[1:] if is_time_dep else target.halo)
```

Comment:
I know this is already in master, but remind me, why not just `halo=target.halo` ?

Reply:
`PETScArray` objects represent `Vec` objects in PETSc, which encapsulate the field at each time step. When using `TimeFunction` objects, I want to exclude the `(0,0)` tuple that appears within the halo.

Reply:
ok, so it's technically more correct to do

```
halo = [target.halo[d] for d in target.space_dimensions]
```

so that you don't depend on the assumption that time is in the 0-th position (you should have used `target._time_position` anyway)